File: agent/deploy/hadesctl/cmd/status.go
/*
Copyright © 2022 chriskali

*/
package cmd

import (
	"fmt"
	"os"
	"os/exec"

	"github.com/nightlyone/lockfile"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

// statusCmd represents the status command
var statusCmd = &cobra.Command{
	Use:   "status",
	Short: "A brief description of your command",
	Run: func(cmd *cobra.Command, args []string) {
		if viper.GetString("service_type") == "systemd" {
			cmd := exec.Command("systemctl", "status", serviceName)
			cmd.Stdout = os.Stdout
			cmd.Stderr = os.Stderr
			cobra.CheckErr(cmd.Run())
		} else if viper.GetString("service_type") == "sysvinit" {
			file, _ := lockfile.New(agentPidFile)
			p, err := file.GetOwner()
			if err != nil {
				fmt.Println("agent is dead")
			} else {
				fmt.Println("agent is running, procs:")
				var pids []int
				var err error
				pids, err = getProcTreeWithCgroup(p.Pid)
				if err != nil {
					pids, _ = getProcTreeWithProc(p.Pid)
				}
				for _, pid := range pids {
					fmt.Println(pid)
				}
			}
		} else if viper.GetString("service_type") == "" {
			fmt.Println("service_type not set")
		} else {
			fmt.Println("unknown service_type: ", viper.GetString("service_type"))
		}
	},
}

func init() {
	rootCmd.AddCommand(statusCmd)
}

File: agent/deploy/hadesctl/src/cmd/status.rs
use anyhow::{bail, Result};
use super::{SERVICE_NAME, collect_proc_tree, read_agent_pid, run_cmd};

pub fn run() -> Result<()> {
    match crate::config::get("service_type").as_str() {
        "systemd" => run_cmd("systemctl", ["status", SERVICE_NAME]),
        "sysvinit" => sysvinit_status(),
        t          => bail!("unknown service_type: '{t}'"),
    }
}

fn sysvinit_status() -> Result<()> {
    match read_agent_pid() {
        Err(_) => {
            println!("hades-agent is dead");
        }
        Ok(pid) => {
            println!("hades-agent is running (pid {pid})");
            let pids = collect_proc_tree(pid);
            if pids.len() > 1 {
                println!("process tree:");
                for p in &pids {
                    println!("  {p}");
                }
            }
        }
    }
    Ok(())
}
